### PR TITLE
Use the flask API to display variables

### LINKED ISSUE #77: Show definition period for variables
As an _analyst_,
I can read the definition period of a variable,
So that I know on which periods I can call it

- - - -

- Consume the data provided by openfisca/openfisca-web-api#125.
- Remove all references to the former API.
- Make the “definition period” header a link to the periods documentation page.

### PR DESCRIPTION
**Pull request by fpagnoux**
Fixes #77

### CODE REVIEW

File: src/components/pages/home.jsx [hunk 1]
```diff
@@ -126,13 +126,12 @@ const SearchResults = React.createClass({
     return (
       <List items={items} type="unstyled">
         {parameterOrVariable => {
-          const {description, itemType, label, name} = parameterOrVariable
-          const displayedDescription = itemType === 'parameter' ? description : label
+          const {description, itemType, name} = parameterOrVariable
```
Comment (MattiSG) [inline src/components/pages/home.jsx]: > `item` would probably be a better name.

File: src/components/variable.jsx [hunk 2]
```diff
@@ -43,275 +15,110 @@ const Variable = React.createClass({
...
-    const {formula, label, source_code} = variable
-    const labelMessage = label || `${variable.name} (aucun label)`
     return (
-      <DocumentTitle title={`${labelMessage} - Explorateur de la législation`}>
+      <DocumentTitle title={`${variable.id} - Explorateur de la législation`}>
         <div>
-          <div className="page-header">
-            <h1>{labelMessage}</h1>
+          <header className="page-header">
+            <h1>{variable.id}</h1>
```
Comment (MattiSG) [inline src/components/variable.jsx]: > `<code>`  (within `<h1>`)? Or style `.page-header h1 { font-family: monospace; }`?

File: src/components/variable.jsx [hunk 3]
```diff
@@ -43,275 +15,110 @@ const Variable = React.createClass({
...
+            <p>{variable.description}</p>
+          </header>
+          {this.renderVariableMetadata(variable)}
+          <div>
+            <ExternalLink href={variable.source}>
+              Voir le code source de cette variable
```
Comment (MattiSG) [inline src/components/variable.jsx]: > “Voir le” not very useful.

File: src/components/variable.jsx [hunk 4]
```diff
@@ -43,275 +15,110 @@ const Variable = React.createClass({
...
+            </ExternalLink>
+          </div>
+          {variable.formulas && this.renderFormulas(variable.formulas)}
+          <div>
+            <ExternalLink href={`${config.apiBaseUrl}/variable/${variable.id}`}>
+              Voir la donnée brute au format JSON
```
Comment (MattiSG) [inline src/components/variable.jsx]: > “Voir la” not very useful.

File: src/components/variable.jsx [hunk 5]
```diff
@@ -43,275 +15,110 @@ const Variable = React.createClass({
...
-        <span>
-          Formule de calcul jusqu'au
-          {" "}
-          <FormattedDate value={formula.stop_instant} />
-        </span>
-      )
+    const definitionPeriodMessage = {
+      YEAR: "de l'année",
+      MONTH: "du mois",
+      ETERNITY: "de l'éternité (cette variable ne peut pas évoluer avec le temps",
```
Comment (MattiSG) [inline src/components/variable.jsx]: Missing closing paren.

File: src/components/variable.jsx [hunk 6]
```diff
@@ -43,275 +15,110 @@ const Variable = React.createClass({
...
-                          }
-                        }
-                      }
-                    </List>
-                  </dd>
-                )
-              }
-            </dl>
+          variable.references && (
+            <span>
```
Comment (MattiSG) [inline src/components/variable.jsx]: An inline `<span>` cannot contain a block-level element `<dt>` / `<dd>`. Within a `<dl>`, only `<dt>` / `<dd>` children are allowed. This `<span>` doesn't seem useful at all, at least for the HTML…?

File: src/components/variable.jsx [hunk 7]
```diff
@@ -43,275 +15,110 @@ const Variable = React.createClass({
...
+              <dt>Références :</dt>
+              <dd><ul>
+                {
+                  variable.references.map((reference, idx) =>
+                    <li key={idx}>
```
Comment (MattiSG) [inline src/components/variable.jsx]: What is this `key` attribute?

File: src/components/variable.jsx [hunk 8]
```diff
@@ -43,275 +15,110 @@ const Variable = React.createClass({
...
-    const {parameters} = this.props
-    const inputVariableNames = formula.input_variables
-    const formulaParameterNames = formula.parameters
     return (
       <div>
+        <dl>
+          <dt>Cette variable est définie pour</dt>
+          <dd>{entityMessage[variable.entity]}.</dd>
+          <dt>Elle se calcule à l'échelle</dt>
+          <dd>{definitionPeriodMessage[variable.definitionPeriod]}</dd>
```
Comment (MattiSG) [inline src/components/variable.jsx]: Inconsistent punctuation: add full stops for each `<dd>`.

File: src/components/variable.jsx [hunk 9]
```diff
@@ -43,275 +15,110 @@ const Variable = React.createClass({
...
+      ETERNITY: "de l'éternité (cette variable ne peut pas évoluer avec le temps",
     }
-  },
-  renderFormula(formula) {
-    const {parameters} = this.props
-    const inputVariableNames = formula.input_variables
-    const formulaParameterNames = formula.parameters
     return (
       <div>
+        <dl>
```
Comment (MattiSG) [inline src/components/variable.jsx]: > Suggested styling (needs a classname): `dt, dd { float: left; } dt { clear: both; }`.

File: src/components/variable.jsx [hunk 10]
```diff
@@ -43,275 +15,110 @@ const Variable = React.createClass({
...
-                        }
-                      }
-                    </List>
-                  </dd>
-                )
-              }
-            </dl>
+          variable.references && (
+            <span>
+              <dt>Références :</dt>
```
Comment (MattiSG) [inline src/components/variable.jsx]: `&nbsp;`

File: src/components/variable.jsx [hunk 11]
```diff
@@ -43,275 +15,110 @@ const Variable = React.createClass({
...
-                  {entry => `${entry[0]} = ${entry[1]}`}
-                </List>
-              </dd>
+                {stopDate && (! startDate) &&
+                  <h3>Jusqu'au <FormattedDate value={stopDate} />&nbsp;:</h3>
+                }
+                {startDate && stopDate &&
+                  <h3>Du <FormattedDate value={startDate} /> au <FormattedDate value={stopDate} />&nbsp;:</h3>
+                }
+                <pre>{formulas[date].content}</pre>
```
Comment (MattiSG) [inline src/components/variable.jsx]: Proper markup would be `<pre><code>`.

File: src/periods.js [hunk 12]
```diff
@@ -0,0 +1,6 @@
+function getDayBefore(day) {
+  const ONE_DAY = 86400000 // in ms
+  return new Date(Date.parse(day) - ONE_DAY).toISOString().slice(0, "YYYY-MM-DD".length)
+}
+
+export default getDayBefore
```
Comment (MattiSG) [inline src/periods.js]: > These modifiers could be added to the function declaration itself.

File: src/webservices.js [hunk 13]
```diff
@@ -4,31 +4,37 @@ import config from "./config"
 
 
 function fetchJson(url, options) {
-  return fetch(url, options).then(parseJsonResponse)
+  return fetch(url, options)
+    .then(response => response.json()
+      .then(data => {
```
Comment (MattiSG) [inline src/webservices.js]: The whole point of promises is that they can be chained. `return` the JSON and use `then` on the return value of `then`, not on `response.json()`.

[See more](http://stackoverflow.com/questions/22944717/callback-pyramid-even-with-when-promise).

File: src/webservices.js [hunk 14]
```diff
@@ -4,31 +4,37 @@ import config from "./config"
...
+        if (response.status >= 200 && response.status < 300) {
+          return {
+            data,
+            'country-package': response.headers.get('country-package'),
+            'country-package-version': response.headers.get('country-package-version'),
+          }
+        } else {
+          throw new Error(JSON.stringify(data.error))
```
Comment (MattiSG) [inline src/webservices.js]: `data.error` might be empty if `response.status` is a 3XX. Need to handle that case.

File: src/components/variable.jsx (hunk 6 above)
Comment (fpagnoux) [inline src/components/variable.jsx]: React imposes to have an "enclosing tag". We can get rid of it by un-factorizing the code I guess.

File: src/components/variable.jsx (hunk 7 above)
Comment (cbenz) [inline src/components/variable.jsx]: This is specific to React: https://facebook.github.io/react/docs/lists-and-keys.html#keys

The key attribute helps re-rendering the DOM in an optimal way:

> Keys help React identify which items have changed, are added, or are removed.

File: src/webservices.js (hunk 13 above)
Comment (fpagnoux) [inline src/webservices.js]: I know but I need the `response` object !

File: src/components/variable.jsx (hunk 2 above)
Comment (fpagnoux) [inline src/components/variable.jsx]: Didn't manage to make it look good. To be done when we redesign the page.

**Pull request merged**